notebooks/Learning_Rate_Scheduler.ipynb

```diff
@@ -0,0 +1,3360 @@
+      "    982        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    983        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    984        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    985        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    986        \u001b[36m0.4469\u001b[0m  0.0000  0.0016\n",
+      "    987        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    988        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    989        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    990        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    991        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    992        \u001b[36m0.4469\u001b[0m  0.0000  0.0016\n",
+      "    993        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    994        \u001b[36m0.4469\u001b[0m  0.0000  0.0018\n",
+      "    995        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    996        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    997        \u001b[36m0.4469\u001b[0m  0.0000  0.0017\n",
+      "    998        0.4469  0.0000  0.0017\n",
+      "    999        0.4469  0.0000  0.0019\n",
+      "   1000        0.4469  0.0000  0.0019\n",
+      "CosineAnnealingLR Score: 0.4500\n",
+      "\n",
+      "Final Results Summary:\n",
+      "\n",
+      "Scheduler: StepLR\n",
+      "Score: 0.4500\n",
+      "Simulated Learning Rates: [0.05 0.05 0.05 0.05 0.05]...\n",
+      "\n",
+      "Scheduler: ReduceLROnPlateau\n",
+      "Score: 0.4000\n",
+      "Simulated Learning Rates: [0.05 0.05 0.05 0.05 0.05]...\n",
+      "\n",
+      "Scheduler: CosineAnnealingLR\n",
+      "Score: 0.4500\n",
+      "Simulated Learning Rates: [0.05       0.04999988 0.04999951 0.04999889 0.04999803]...\n"
+     ]
+    }
+   ],
+   "source": [
+    "def simulate_scheduler(scheduler_class, optimizer_params, steps, initial_lr):\n",
+    "    optimizer = torch.optim.SGD([torch.zeros(1, requires_grad=True)], lr=initial_lr)\n",
+    "    scheduler = scheduler_class(optimizer, **optimizer_params)\n",
+    "    lrs = []\n",
+    "\n",
+    "    dummy_metric = 0.5 \n",
+    "    for _ in range(steps):\n",
+    "        lrs.append(optimizer.param_groups[0]['lr'])\n",
+    "\n",
+    "        if isinstance(scheduler, torch.optim.lr_scheduler.ReduceLROnPlateau):\n",
+    "            scheduler.step(dummy_metric)\n",
+    "        else:\n",
+    "            scheduler.step()\n",
+    "\n",
+    "    return np.array(lrs)\n",
+    "\n",
+    "\n",
+    "\n",
+    "def train_schedulers(X_train, X_test, y_train, y_test, lr=0.05, epochs=1000, hidden_dim=128):\n",
+    "    X_train = torch.tensor(X_train, dtype=torch.float32)\n",
+    "    X_test = torch.tensor(X_test, dtype=torch.float32)\n",
+    "    y_train = torch.tensor(y_train, dtype=torch.float32).view(-1, 1)\n",
+    "    y_test = torch.tensor(y_test, dtype=torch.float32).view(-1, 1)\n",
```

Comment:
I kept getting the following error.

RuntimeError: mat1 and mat2 must have the same dtype, but got Double and Float

So, I have kept both the tensors in float 32 to avoid the mismatch. If it a necessary change, I will modify it.

Reply:
Updated as discussed.